GitCommands/ArgumentBuilderExtensions.cs

```diff
@@ -300,5 +301,92 @@ public static void Add(this ArgumentBuilder builder, [CanBeNull, ItemCanBeNull]
+        {
+            // 3: double quotes + ' '
+            // '"git.exe" ' is always included in final command line arguments
+            if (!baseLength.HasValue)
+            {
+                baseLength = AppSettings.GitCommand.Length + 3;
+            }
+
+            var baseCommand = builder.ToString();
+            if (baseLength + baseCommand.Length >= maxLength)
+            {
+                throw new ArgumentException($"Git base command \"{baseCommand}\" always reached max length of {maxLength} characters.", nameof(baseCommand));
+            }
+
+            // Clone command as argument builder
+            var tmpBuilder = new ArgumentBuilder() { baseCommand };
+            var result = new List<BatchArgumentItem>();
+            var itemCount = 0;
+
+            var argBuilder = arguments.Aggregate(tmpBuilder, (curr, value) =>
+            {
+                // 1: ' ' space character length will be added
+                // When enumeration is finished, no need to add ' ' to length calculcation
+                if (baseLength + curr.Length + 1 + value.Length >= maxLength)
+                {
+                    // Handle abnormal case when base command and a single argument exceed max length
+                    if (itemCount == 0)
+                    {
+                        throw new ArgumentException($"Git command \"{curr}\" always exceeded max length of {maxLength} characters.", nameof(arguments));
+                    }
+
+                    // Finish current command line
+                    result.Add(new BatchArgumentItem
+                    {
+                        Argument = curr.ToString(),
+                        Count = itemCount
+                    });
+
+                    // Return new argument builder
+                    itemCount = 1;
+                    return new ArgumentBuilder() { baseCommand, value };
+                }
+
+                curr.Add(value);
+                itemCount++;
+                return curr;
+            });
+
+            // Handle rare case when last argument length exceed max length
+            if (baseLength + argBuilder.Length >= maxLength)
+            {
+                throw new ArgumentException($"Git command \"{argBuilder}\" always exceeded max length of {maxLength} characters.", nameof(arguments));
+            }
+
+            // Add last commandline batch
+            if (!argBuilder.IsEmpty)
+            {
+                result.Add(new BatchArgumentItem
+                {
+                    Argument = argBuilder.ToString(),
+                    Count = itemCount
+                });
+            }
+
+            return result;
+        }
+    }
+
+    public sealed class BatchArgumentItem
+    {
+        public ArgumentString Argument { get; set; }
+
+        public int Count { get; set; }
```

Comment:
Consider making these get-only and adding a constructor.

Reply:
Done